### PR TITLE
feat: user-provided ClusterRole rules for Eirini

### LINKED ISSUE #595: Allow the user to parameterized ClusterRole rules resources for Eirini
**Is your feature request related to a problem? Please describe.**
Yes, there is no way to extend the resources rules in https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/templates/eirini.yaml#L12 , unless you apply a manual change or a patch of the file. For some IaaS, the resources list requires more items.

**Describe the solution you'd like**
A feature flag in the values.yml

**Describe alternatives you've considered**
None

**Additional context**
Yes, in IKS, we required to have the following resources:
```
resources: ['mutatingwebhookconfigurations', 'nodes', 'nodes/proxy']
```

### PR DESCRIPTION
**Pull request by EmilyEmily**
Fixes #595.

## Description
1. Added eirini cluster-role resources and verb properties under feature eirini. 
2. Add if condition in values.yaml, set customized resources and verb if eirini enabled and there are customized values for them. 

## Motivation and Context
Since there is no way to extend the resources rules in https://github.com/cloudfoundry-incubator/kubecf/blob/master/deploy/helm/kubecf/templates/eirini.yaml#L12 , unless you apply a manual change or a patch of the file. For some IaaS, the resources list requires more items, for example in IKS cluster, it needs this resource list: 
```
resources: ['mutatingwebhookconfigurations', 'nodes', 'nodes/proxy']
```

## How Has This Been Tested?
I did two tests, one is only to enable eirini feature, no properties under it. 
The other is to enable eirini feature, and added customized resources under it. 
Both deployment succeeded. 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] My code has security implications.
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: deploy/helm/kubecf/values.yaml [hunk 1]
```diff
@@ -238,6 +238,12 @@ features:
     registry:
       service:
         nodePort: 32123
+    rbac:
+      cluster_role:
+        eirini_webhook:
```
Comment (f0rmiga) [inline deploy/helm/kubecf/values.yaml:243]: +1 on being specific here on the `eirini_webhook`. I'd swap it with the `cluster_role` level:

```yaml
rbac:
  eirini_webhook:
    cluster_role:
      ...
```

File: deploy/helm/kubecf/values.yaml [hunk 2]
```diff
@@ -238,6 +238,12 @@ features:
...
+          rules:
+            resources: ~
+            verbs: ~
```
Comment (f0rmiga) [inline deploy/helm/kubecf/values.yaml:246]: As I suggested on https://github.com/cloudfoundry-incubator/kubecf/issues/595, you can use `rules` as an array. The goal is to match what the user provides here and what the Kubernetes API expects.

```suggestion
          rules: []
```

File: deploy/helm/kubecf/templates/eirini.yaml [hunk 3]
```diff
@@ -11,8 +11,17 @@ metadata:
...
+{{- else }}
   resources: ['mutatingwebhookconfigurations']
-  verbs:     ['*']
+{{- end }}
+
+{{- if .Values.features.eirini.rbac.cluster_role.eirini_webhook.rules.verbs }}
+  verbs: {{ toJson .Values.features.eirini.rbac.cluster_role.eirini_webhook.rules.verbs }}
+{{- else }}
+  verbs: ['*']
+{{- end }}
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/eirini.yaml:24]: As suggested on https://github.com/cloudfoundry-incubator/kubecf/issues/595, we don't want to mix the user-provided  ClusterRole rules with the static ones. The suggestion was to add an _extra_ ClusterRole, and bind it to the eirini ServiceAccount. This way it's clearer during runtime what the user provided and what KubeCF provided by default.

File: deploy/helm/kubecf/values.yaml (hunk 1 above)
Comment (EmilyEmily) [reply deploy/helm/kubecf/values.yaml:243]: This is a good idea. Thanks, change it to: 
```
cluster_role:
  rules: 
......
```

File: deploy/helm/kubecf/values.yaml (hunk 2 above)
Comment (EmilyEmily) [reply deploy/helm/kubecf/values.yaml:246]: Thanks, changed.

File: deploy/helm/kubecf/templates/eirini.yaml (hunk 3 above)
Comment (EmilyEmily) [reply deploy/helm/kubecf/templates/eirini.yaml:24]: Got it. Change the ClusterRole name to eirini_webhook_user_provided.

**Comment by EmilyEmily** [comment]
Verified on my own pipeline: https://concourse.cf.w3.cloud.ibm.com/teams/gen2dev/pipelines/bkcf-emily

See job https://concourse.cf.w3.cloud.ibm.com/teams/gen2dev/pipelines/bkcf-emily/jobs/deploy-previous-diego/builds/3
https://concourse.cf.w3.cloud.ibm.com/teams/gen2dev/pipelines/bkcf-emily/jobs/upgrade-latest-diego/builds/5
also
https://concourse.cf.w3.cloud.ibm.com/teams/gen2dev/pipelines/bkcf-emily/jobs/deploy-charts-diego/builds/32

**Comment by EmilyEmily** [comment]
@f0rmiga thank you for the clear review. I've fixed those issues. Could you help review again? Thank you.

File: deploy/helm/kubecf/values.yaml [hunk 4]
```diff
@@ -238,6 +238,8 @@ features:
     registry:
       service:
         nodePort: 32123
+    cluster_role:
+          rules: []
```
Comment (f0rmiga) [inline deploy/helm/kubecf/values.yaml:242]: It's missing the keys from before.

```suggestion
    rbac:
      webhook:
        cluster_role:
          rules: []
```

File: deploy/helm/kubecf/templates/eirini.yaml [hunk 5]
```diff
@@ -28,6 +28,31 @@ subjects:
   name: eirini
   namespace: {{ .Release.Namespace }}
 
+{{- if gt (len .Values.features.eirini.cluster_role.rules) 0 }}
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/eirini.yaml:31]:
```suggestion
{{- if gt (len .Values.features.eirini.rbac.webhook.cluster_role.rules) 0 }}
```

File: deploy/helm/kubecf/templates/eirini.yaml [hunk 6]
```diff
@@ -28,6 +28,31 @@ subjects:
...
+# eirini-webhook-user-provided cluster role for eirini service account.
+# If user provides a customized cluster role, it is used to implement eirinix extensions.
+---
+apiVersion: rbac.authorization.k8s.io/v1
+kind: ClusterRole
+metadata:
+  name: {{ .Release.Namespace }}:eirini-webhook-user-provided
+rules: {{ .Values.features.eirini.cluster_role.rules | toJson }}
```
Comment (f0rmiga) [inline deploy/helm/kubecf/templates/eirini.yaml:39]:
```suggestion
rules: {{ .Values.features.eirini.rbac.webhook.cluster_role.rules | toJson }}
```

**Pull request merged**